File: ios/NaSumShuttle/AppShell/AppModel.swift
import AuthenticationServices
import CoreLocation
import Foundation
import Observation
import SwiftUI
import UIKit

@MainActor
@Observable
final class AppModel {
    enum Mode {
        case live
        case previewLoggedOut
        case previewLoggedIn
    }

    struct NotificationNavigationTarget: Equatable {
        let notificationId: String?
        let routeCode: String
        let routeStopId: String?
    }

    struct ScanNavigationTarget: Equatable {
        let routeCode: String
    }

    private enum Constants {
        static let keychainService = "org.nasumik.NaSumShuttle"
        static let sessionAccount = "session-jwt"
        static let themePreferenceKey = "org.nasumik.NaSumShuttle.themePreference"
    }

    let apiClient: APIClient
    let authProvider: AuthProviding
    let googleAuthProvider: GoogleAuthManager
    let pushManager: PushNotificationManager
    let adminStore: AdminStore

    var isBootstrapping = true
    var isLoading = false
    var isAuthenticating = false
    var errorMessage: String?

    var sessionToken: String?
    var currentUser: MeResponse?
    var routeSummaries: [RouteSummary] = []
    var registration: RegistrationEnvelope?
    var places: [PlaceSummary] = []
    var notifications: [AppNotification] = []
    var unreadCount = 0
    var selectedRouteCode: String?
    var pendingNotificationNavigation: NotificationNavigationTarget?
    var pendingScanNavigation: ScanNavigationTarget?
    var routeDetails: [String: RouteDetail] = [:]
    var routeCandidates: [String: PlaceRoutesResponse] = [:]
    var runInfoByRouteCode: [String: CheckInRunInfoResponse] = [:]
    var lastCheckInResponse: CheckInResponse?
    var themePreference: AppThemePreference = .system {
        didSet {
            UserDefaults.standard.set(themePreference.rawValue, forKey: Constants.themePreferenceKey)
        }
    }

    private let mode: Mode

    private static var isRunningForPreviews: Bool {
        ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1"
    }

    init(mode: Mode = .live) {
        let effectiveMode: Mode = Self.isRunningForPreviews && mode == .live ? .previewLoggedOut : mode
        self.mode = effectiveMode
        self.apiClient = APIClient()
        self.authProvider = effectiveMode == .live ? LineAuthManager() : PreviewAuthProvider()
        self.googleAuthProvider = GoogleAuthManager()
        self.pushManager = PushNotificationManager()
        self.adminStore = AdminStore(apiClient: apiClient)
        self.themePreference = Self.loadThemePreference()

        apiClient.sessionTokenProvider = { [weak self] in
            self?.sessionToken
        }
        apiClient.unauthorizedHandler = { [weak self] in
            Task { @MainActor in
                await self?.handleUnauthorized()
            }
        }
        pushManager.onDeviceTokenUpdated = { [weak self] token in
            Task { @MainActor in
                await self?.registerPushTokenIfPossible(token: token)
            }
        }
        pushManager.onNotificationTapped = { [weak self] payload in
            Task { @MainActor in
                await self?.handleNotificationTap(payload)
            }
        }

        if effectiveMode == .previewLoggedIn {
            applyPreviewState()
        }
        if effectiveMode != .live {
            isBootstrapping = false
        }
    }

    var isAuthenticated: Bool {
        sessionToken != nil && currentUser != nil
    }

    var isInitialDataLoading: Bool {
        isBootstrapping || (isLoading && currentUser == nil)
    }

    var errorAlertTitle: String {
        guard errorMessage == APIError.unauthorized.localizedDescription else {
            return RiderStrings.commonServerError(preferredLanguage)
        }
        return preferredLanguage == .ko ? "세션이 만료되었습니다." : "Session Expired"
    }

    var preferredLanguage: AppLanguage {
        currentUser?.preferredLanguage ?? .ko
    }

    var preferredColorScheme: ColorScheme? {
        switch themePreference {
        case .system:
            return nil
        case .light:
            return .light
        case .dark:
            return .dark
        }
    }

    var isAdminSurfaceEnabled: Bool {
        guard let role = currentUser?.role else { return false }
        return role == .admin || role == .driver
    }

    private static func loadThemePreference() -> AppThemePreference {
        guard
            let rawValue = UserDefaults.standard.string(forKey: Constants.themePreferenceKey),
            let preference = AppThemePreference(rawValue: rawValue)
        else {
            return .system
        }
        return preference
    }

    func bootstrap() async {
        defer { isBootstrapping = false }

        switch mode {
        case .previewLoggedOut:
            return
        case .previewLoggedIn:
            return
        case .live:
            break
        }

        sessionToken = SecureStore.load(service: Constants.keychainService, account: Constants.sessionAccount)
        guard sessionToken != nil else { return }

        do {
            try await refreshAll()
            await pushManager.refreshAuthorizationStatus()
            await registerPushTokenIfPossible(token: pushManager.deviceTokenHex)
        } catch {
            await handleUnauthorized()
        }
    }

    func signIn(presentingViewController: UIViewController?) async {
        guard mode == .live else { return }
        isAuthenticating = true
        errorMessage = nil

        defer { isAuthenticating = false }

        do {
            let identity = try await authProvider.login(presentingViewController: presentingViewController)
            let session = try await apiClient.exchangeLineSession(accessToken: identity.accessToken)

            try await applySession(session)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func signInWithApple(authorization: ASAuthorization, nonce: String) async {
        guard mode == .live else { return }
        isAuthenticating = true
        errorMessage = nil

        defer { isAuthenticating = false }

        do {
            let credential = try AppleAuthParser.credential(from: authorization, nonce: nonce)
            let session = try await apiClient.exchangeAppleSession(
                identityToken: credential.identityToken,
                authorizationCode: credential.authorizationCode,
                nonce: credential.nonce
            )

            try await applySession(session)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func signInWithGoogle(presentingViewController: UIViewController?) async {
        guard mode == .live else { return }
        isAuthenticating = true
        errorMessage = nil

        defer { isAuthenticating = false }

        do {
            let credential = try await googleAuthProvider.signIn(presentingViewController: presentingViewController)
            let session = try await apiClient.exchangeGoogleSession(idToken: credential.idToken)

            try await applySession(session)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func signInWithEmail(email: String, password: String) async {
        guard mode == .live else { return }
        isAuthenticating = true
        errorMessage = nil

        defer { isAuthenticating = false }

        do {
            let session = try await apiClient.exchangeEmailPasswordSession(
                email: email,
                password: password
            )

            try await applySession(session)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func logout() async {
        if let token = pushManager.deviceTokenHex {
            _ = try? await apiClient.deletePushToken(token: token)
        }

        await authProvider.logout()
        googleAuthProvider.logout()
        sessionToken = nil
        currentUser = nil
        registration = nil
        routeSummaries = []
        places = []
        notifications = []
        unreadCount = 0
        selectedRouteCode = nil
        pendingNotificationNavigation = nil
        pendingScanNavigation = nil
        routeDetails = [:]
        runInfoByRouteCode = [:]
        SecureStore.delete(service: Constants.keychainService, account: Constants.sessionAccount)
    }

    func refreshAll() async throws {
        guard mode == .live else { return }

        isLoading = true
        defer { isLoading = false }

        async let meTask = apiClient.fetchMe()
        async let routesTask = apiClient.fetchRouteSummaries()
        async let registrationTask = apiClient.fetchRegistration()
        async let placesTask = apiClient.fetchPlaces()
        async let notificationsTask = apiClient.fetchNotifications()
        async let unreadTask = apiClient.fetchUnreadCount()

        let me = try await meTask
        let routes = try await routesTask
        let registration = try await registrationTask
        let places = try await placesTask
        let notifications = try await notificationsTask
        let unreadCount = try await unreadTask

        currentUser = me
        routeSummaries = routes
        self.registration = registration
        self.places = places
        self.notifications = notifications
        self.unreadCount = unreadCount.unreadCount

        let preferredRouteCode = registration.registration?.route.routeCode ?? routes.first?.routeCode
        if let preferredRouteCode {
            selectedRouteCode = preferredRouteCode
            try await loadRouteDetail(routeCode: preferredRouteCode)
            try? await loadRunInfo(routeCode: preferredRouteCode)
        }
    }

    func loadRouteDetail(routeCode: String) async throws {
        if routeDetails[routeCode] != nil { return }
        let detail = try await apiClient.fetchRouteDetail(routeCode: routeCode)
        routeDetails[routeCode] = detail
    }

    func selectRoute(routeCode: String) async {
        selectedRouteCode = routeCode
        do {
            try await loadRouteDetail(routeCode: routeCode)
            try? await loadRunInfo(routeCode: routeCode)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func loadPlaceRoutes(googlePlaceId: String) async {
        if routeCandidates[googlePlaceId] != nil { return }
        do {
            routeCandidates[googlePlaceId] = try await apiClient.fetchPlaceRoutes(googlePlaceId: googlePlaceId)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func registerStop(routeCode: String, routeStopId: String) async {
        do {
            _ = try await apiClient.updateRegistration(routeCode: routeCode, routeStopId: routeStopId)
            registration = try await apiClient.fetchRegistration()
            if let routeCode = registration?.registration?.route.routeCode {
                selectedRouteCode = routeCode
                routeDetails[routeCode] = registration?.registration?.route
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func removeRegistration() async {
        do {
            _ = try await apiClient.deleteRegistration()
            registration = try await apiClient.fetchRegistration()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func reloadNotifications() async {
        do {
            notifications = try await apiClient.fetchNotifications()
            let unreadResponse = try await apiClient.fetchUnreadCount()
            unreadCount = unreadResponse.unreadCount
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func markNotificationRead(_ notification: AppNotification) async {
        do {
            _ = try await apiClient.markNotificationRead(id: notification.id)
            await reloadNotifications()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func markAllNotificationsRead() async {
        do {
            _ = try await apiClient.markAllNotificationsRead()
            await reloadNotifications()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updatePreferences(pushNotificationsEnabled: Bool? = nil, preferredLanguage: AppLanguage? = nil) async {
        do {
            _ = try await apiClient.updatePreferences(
                pushNotificationsEnabled: pushNotificationsEnabled,
                preferredLanguage: preferredLanguage
            )
            currentUser = try await apiClient.fetchMe()
            if pushNotificationsEnabled == true {
                await pushManager.requestAuthorizationIfNeeded()
            }
            if pushNotificationsEnabled == false, let token = pushManager.deviceTokenHex {
                _ = try? await apiClient.deletePushToken(token: token)
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func loadRunInfo(routeCode: String) async throws {
        runInfoByRouteCode[routeCode] = try await apiClient.fetchRunInfo(routeCode: routeCode)
    }

    func checkIn(routeCode: String, routeStopId: String, additionalPassengers: Int) async {
        do {
            if runInfoByRouteCode[routeCode] == nil {
                try await loadRunInfo(routeCode: routeCode)
            }
            guard let run = runInfoByRouteCode[routeCode]?.run else { return }
            lastCheckInResponse = try await apiClient.checkIn(
                runId: run.id,
                routeStopId: routeStopId,
                additionalPassengers: additionalPassengers
            )
            try await loadRunInfo(routeCode: routeCode)
            await reloadNotifications()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func parseRouteCode(from scannedText: String) -> String? {
        let trimmed = scannedText.trimmingCharacters(in: .whitespacesAndNewlines)
        if let url = URL(string: trimmed), let routeCode = routeCode(from: url) {
            return routeCode
        }

        return trimmed.isEmpty ? nil : trimmed
    }

    private func routeCode(from url: URL) -> String? {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return nil
        }

        if let routeCode = components.queryItems?.first(where: { $0.name == "routeCode" })?.value {
            return routeCode
        }

        if
            let liffState = components.queryItems?.first(where: { $0.name == "liff.state" })?.value,
            let routeCode = routeCodeFromNestedURL(liffState)
        {
            return routeCode
        }

        if
            let sessionParams = components.queryItems?.first(where: { $0.name == "sessionParams" })?.value,
            let routeCode = routeCodeFromSessionParams(sessionParams)
        {
            return routeCode
        }

        return nil
    }

    private func routeCodeFromNestedURL(_ value: String) -> String? {
        if let url = URL(string: value), let routeCode = routeCode(from: url) {
            return routeCode
        }

        guard
            let base = URL(string: "https://dummy.local"),
            let relativeURL = URL(string: value, relativeTo: base)?.absoluteURL
        else {
            return nil
        }

        return routeCode(from: relativeURL)
    }

    private func routeCodeFromSessionParams(_ value: String) -> String? {
        guard
            let data = value.data(using: .utf8),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let routeCode = json["routeCode"] as? String
        else {
            return nil
        }

        return routeCode
    }

    func clearError() {
        errorMessage = nil
    }

    func handleIncomingURL(_ url: URL) -> Bool {
        if googleAuthProvider.handleOpenURL(url) || authProvider.handleOpenURL(url) {
            return true
        }

        guard
            isUniversalScanURL(url),
            let routeCode = parseRouteCode(from: url.absoluteString)
        else {
            return false
        }

        pendingScanNavigation = ScanNavigationTarget(routeCode: routeCode)
        selectedRouteCode = routeCode
        return true
    }

    private func isUniversalScanURL(_ url: URL) -> Bool {
        guard
            let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
            components.scheme == "https",
            components.host == "nasum-church-shuttle.vercel.app"
        else {
            return false
        }
        return components.path == "/scan"
    }

    func registerPushTokenIfPossible(token: String?) async {
        guard
            mode == .live,
            let token,
            currentUser?.pushNotificationsEnabled == true
        else {
            return
        }

        do {
            _ = try await apiClient.registerPushToken(
                token: token,
                bundleId: AppConfiguration.apnsBundleID,
                environment: AppConfiguration.apnsEnvironment
            )
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func handleNotificationTap(_ payload: PushNotificationTapPayload) async {
        guard mode == .live else { return }

        do {
            if let notificationId = payload.notificationId {
                _ = try? await apiClient.markNotificationRead(id: notificationId)
            }
            try? await reloadNotificationsAfterPushTap()

            guard let routeCode = payload.routeCode else { return }
            selectedRouteCode = routeCode
            try await loadRouteDetail(routeCode: routeCode)
            try? await loadRunInfo(routeCode: routeCode)
            pendingNotificationNavigation = NotificationNavigationTarget(
                notificationId: payload.notificationId,
                routeCode: routeCode,
                routeStopId: payload.userRouteStopId
            )
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func handleUnauthorized() async {
        await logout()
        errorMessage = APIError.unauthorized.localizedDescription
    }

    private func reloadNotificationsAfterPushTap() async throws {
        notifications = try await apiClient.fetchNotifications()
        unreadCount = try await apiClient.fetchUnreadCount().unreadCount
    }

    private func applySession(_ session: SessionExchangeResponse) async throws {
        sessionToken = session.sessionToken
        SecureStore.save(session.sessionToken, service: Constants.keychainService, account: Constants.sessionAccount)
        try await refreshAll()
        await pushManager.requestAuthorizationIfNeeded()
        await registerPushTokenIfPossible(token: pushManager.deviceTokenHex)
    }

    private func applyPreviewState() {
        sessionToken = "preview"
        currentUser = PreviewFixtures.me
        routeSummaries = PreviewFixtures.routeSummaries
        registration = PreviewFixtures.registration
        routeDetails[PreviewFixtures.routeDetail.routeCode] = PreviewFixtures.routeDetail
        selectedRouteCode = PreviewFixtures.routeDetail.routeCode
        places = [
            PlaceSummary(
                googlePlaceId: "google-1",
                name: "강남역",
                lat: 37.4979,
                lng: 127.0276,
                isTerminal: false
            )
        ]
        notifications = PreviewFixtures.notifications
        unreadCount = notifications.filter { !$0.isRead }.count
        runInfoByRouteCode[PreviewFixtures.routeDetail.routeCode] = CheckInRunInfoResponse(
            run: ShuttleRun(
                id: "run-1",
                routeId: PreviewFixtures.routeDetail.id,
                serviceDate: ISO8601DateFormatter().string(from: .now),
                status: "active",
                startedAt: ISO8601DateFormatter().string(from: .now),
                endedAt: nil
            ),
            route: PreviewFixtures.routeDetail,
            stopStates: [
                StopBoardingState(routeStopId: "stop-1", totalPassengers: 12, status: "arrived")
            ],
            myCheckin: nil
        )
        isBootstrapping = false
    }
}

File: android/app/src/main/java/sg/nasumchurch/shuttle/app/AppModel.kt
package sg.nasumchurch.shuttle.app

import android.app.Application
import android.net.Uri
import java.net.URLDecoder
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import sg.nasumchurch.shuttle.auth.SessionStore
import sg.nasumchurch.shuttle.network.APIClient
import sg.nasumchurch.shuttle.preview.PreviewFixtures
import sg.nasumchurch.shuttle.sharedmodel.*

data class NotificationNavigationTarget(
    val notificationId: String?,
    val routeCode: String,
    val routeStopId: String?,
)

data class ScanNavigationTarget(val routeCode: String)

data class AppUiState(
    val isBootstrapping: Boolean = true,
    val isLoading: Boolean = false,
    val isAuthenticating: Boolean = false,
    val errorMessage: String? = null,
    val sessionToken: String? = null,
    val currentUser: MeResponse? = null,
    val routeSummaries: List<RouteSummary> = emptyList(),
    val registration: RegistrationEnvelope? = null,
    val places: List<PlaceSummary> = emptyList(),
    val notifications: List<AppNotification> = emptyList(),
    val unreadCount: Int = 0,
    val selectedRouteCode: String? = null,
    val routeDetails: Map<String, RouteDetail> = emptyMap(),
    val routeCandidates: Map<String, PlaceRoutesResponse> = emptyMap(),
    val runInfoByRouteCode: Map<String, CheckInRunInfoResponse> = emptyMap(),
    val pendingNotificationNavigation: NotificationNavigationTarget? = null,
    val pendingScanNavigation: ScanNavigationTarget? = null,
    val adminRuns: List<AdminRun> = emptyList(),
    val adminRegistrations: List<AdminRegistrationRow> = emptyList(),
    val adminUsers: List<AdminPrivilegedUser> = emptyList(),
    val themePreference: AppThemePreference = AppThemePreference.system,
) {
    val isAuthenticated: Boolean get() = sessionToken != null && currentUser != null
    val preferredLanguage: AppLanguage get() = currentUser?.preferredLanguage ?: AppLanguage.ko
    val isAdminSurfaceEnabled: Boolean get() = currentUser?.role == UserRole.admin || currentUser?.role == UserRole.driver
    val selectedRoute: RouteDetail? get() = selectedRouteCode?.let { routeDetails[it] }
}

class AppModel(application: Application) : AndroidViewModel(application) {
    private val sessionStore = SessionStore(application)
    private val apiClient = APIClient(sessionTokenProvider = { state.value.sessionToken })
    private val _state = MutableStateFlow(AppUiState())
    val state: StateFlow<AppUiState> = _state

    fun bootstrap(initialUri: Uri? = null) {
        viewModelScope.launch {
            val token = sessionStore.loadSessionToken()
            val theme = sessionStore.loadTheme()?.let { runCatching { AppThemePreference.valueOf(it) }.getOrNull() }
                ?: AppThemePreference.system
            _state.update { it.copy(sessionToken = token, themePreference = theme) }
            initialUri?.let { handleIncomingUri(it) }
            if (token == null) {
                _state.update { it.copy(isBootstrapping = false) }
                return@launch
            }
            runCatching { refreshAll() }
                .onFailure { logout(error = it.message ?: "Session expired.") }
            _state.update { it.copy(isBootstrapping = false) }
        }
    }

    fun usePreviewState() {
        _state.value = AppUiState(
            isBootstrapping = false,
            sessionToken = "preview",
            currentUser = PreviewFixtures.me,
            routeSummaries = PreviewFixtures.routeSummaries,
            registration = PreviewFixtures.registration,
            places = PreviewFixtures.places,
            notifications = PreviewFixtures.notifications,
            unreadCount = PreviewFixtures.notifications.count { !it.isRead },
            selectedRouteCode = PreviewFixtures.routeDetail.routeCode,
            routeDetails = mapOf(PreviewFixtures.routeDetail.routeCode to PreviewFixtures.routeDetail),
            adminRuns = PreviewFixtures.adminRuns,
            adminRegistrations = PreviewFixtures.adminRegistrations,
            adminUsers = PreviewFixtures.adminUsers,
        )
    }

    fun signInWithEmail(email: String, password: String) {
        viewModelScope.launch {
            _state.update { it.copy(isAuthenticating = true, errorMessage = null) }
            runCatching {
                val session = apiClient.exchangeEmailPasswordSession(email, password)
                applySession(session.token)
            }.onFailure { caught ->
                _state.update { it.copy(errorMessage = caught.message ?: "Sign-in failed.") }
            }
            _state.update { it.copy(isAuthenticating = false) }
        }
    }

    fun unsupportedProvider(provider: String) {
        _state.update {
            it.copy(errorMessage = "$provider login is scaffolded. Add provider SDK credentials in Android Studio to enable it.")
        }
    }

    fun refreshAllAsync() {
        viewModelScope.launch {
            refreshAll()
        }
    }

    private suspend fun refreshAll() {
        _state.update { it.copy(isLoading = true, errorMessage = null) }
        val me = apiClient.fetchMe()
        val routes = apiClient.fetchRouteSummaries()
        val registration = apiClient.fetchRegistration()
        val places = apiClient.fetchPlaces()
        val notifications = apiClient.fetchNotifications()
        val unread = apiClient.fetchUnreadCount()
        val preferredRouteCode = registration.registration?.route?.routeCode ?: routes.firstOrNull()?.routeCode
        val details = mutableMapOf<String, RouteDetail>()
        preferredRouteCode?.let { details[it] = apiClient.fetchRouteDetail(it) }
        _state.update {
            it.copy(
                isLoading = false,
                currentUser = me,
                routeSummaries = routes,
                registration = registration,
                places = places,
                notifications = notifications,
                unreadCount = unread.unreadCount,
                selectedRouteCode = preferredRouteCode,
                routeDetails = it.routeDetails + details,
            )
        }
    }

    fun selectRoute(routeCode: String) {
        viewModelScope.launch {
            runCatching {
                val detail = state.value.routeDetails[routeCode] ?: apiClient.fetchRouteDetail(routeCode)
                val runInfo = runCatching { apiClient.fetchRunInfo(routeCode) }.getOrNull()
                _state.update {
                    it.copy(
                        selectedRouteCode = routeCode,
                        routeDetails = it.routeDetails + (routeCode to detail),
                        runInfoByRouteCode = if (runInfo == null) it.runInfoByRouteCode else it.runInfoByRouteCode + (routeCode to runInfo),
                    )
                }
            }.onFailure { setError(it) }
        }
    }

    fun loadPlaceRoutes(place: PlaceSummary) {
        viewModelScope.launch {
            if (state.value.routeCandidates.containsKey(place.googlePlaceId)) return@launch
            runCatching { apiClient.fetchPlaceRoutes(place.googlePlaceId) }
                .onSuccess { response ->
                    _state.update { it.copy(routeCandidates = it.routeCandidates + (place.googlePlaceId to response)) }
                }
                .onFailure { setError(it) }
        }
    }

    fun registerStop(routeCode: String, routeStopId: String) {
        viewModelScope.launch {
            runCatching {
                apiClient.updateRegistration(routeCode, routeStopId)
                val registration = apiClient.fetchRegistration()
                _state.update {
                    it.copy(
                        registration = registration,
                        selectedRouteCode = registration.registration?.route?.routeCode ?: routeCode,
                    )
                }
            }.onFailure { setError(it) }
        }
    }

    fun checkIn(routeCode: String, routeStopId: String, additionalPassengers: Int) {
        viewModelScope.launch {
            runCatching {
                val runInfo = state.value.runInfoByRouteCode[routeCode] ?: apiClient.fetchRunInfo(routeCode)
                val run = runInfo.run ?: throw IllegalStateException("No active run found for this route.")
                apiClient.checkIn(run.id, routeStopId, additionalPassengers)
                val updatedRunInfo = apiClient.fetchRunInfo(routeCode)
                _state.update { it.copy(runInfoByRouteCode = it.runInfoByRouteCode + (routeCode to updatedRunInfo)) }
                reloadNotifications()
            }.onFailure { setError(it) }
        }
    }

    fun reloadNotifications() {
        viewModelScope.launch {
            runCatching {
                val notifications = apiClient.fetchNotifications()
                val unread = apiClient.fetchUnreadCount()
                _state.update { it.copy(notifications = notifications, unreadCount = unread.unreadCount) }
            }.onFailure { setError(it) }
        }
    }

    fun markNotificationRead(notification: AppNotification) {
        viewModelScope.launch {
            runCatching {
                apiClient.markNotificationRead(notification.id)
                reloadNotifications()
            }.onFailure { setError(it) }
        }
    }

    fun updatePreferences(pushNotificationsEnabled: Boolean? = null, preferredLanguage: AppLanguage? = null, theme: AppThemePreference? = null) {
        viewModelScope.launch {
            runCatching {
                if (theme != null) {
                    sessionStore.saveTheme(theme.name)
                    _state.update { it.copy(themePreference = theme) }
                }
                if (pushNotificationsEnabled != null || preferredLanguage != null) {
                    apiClient.updatePreferences(pushNotificationsEnabled, preferredLanguage)
                    _state.update { it.copy(currentUser = apiClient.fetchMe()) }
                }
            }.onFailure { setError(it) }
        }
    }

    fun registerPushToken(token: String) {
        viewModelScope.launch {
            if (state.value.currentUser?.pushNotificationsEnabled != true) return@launch
            runCatching { apiClient.registerAndroidPushToken(token) }.onFailure { setError(it) }
        }
    }

    fun loadAdmin() {
        viewModelScope.launch {
            runCatching {
                val runs = apiClient.fetchAdminRuns("active")
                val registrations = if (state.value.currentUser?.role == UserRole.admin) apiClient.fetchAdminRegistrations() else emptyList()
                val users = if (state.value.currentUser?.role == UserRole.admin) apiClient.fetchAdminUsers() else emptyList()
                _state.update { it.copy(adminRuns = runs, adminRegistrations = registrations, adminUsers = users) }
            }.onFailure { setError(it) }
        }
    }

    fun handleIncomingUri(uri: Uri): Boolean {
        val routeCode = routeCodeFromUri(uri) ?: return false
        _state.update {
            it.copy(
                pendingScanNavigation = ScanNavigationTarget(routeCode),
                selectedRouteCode = routeCode,
            )
        }
        return true
    }

    fun consumeScanNavigation() {
        _state.update { it.copy(pendingScanNavigation = null) }
    }

    fun handleNotificationPayload(data: Map<String, String>) {
        val routeCode = data["routeCode"] ?: return
        _state.update {
            it.copy(
                pendingNotificationNavigation = NotificationNavigationTarget(
                    notificationId = data["notificationId"],
                    routeCode = routeCode,
                    routeStopId = data["userRouteStopId"],
                ),
                selectedRouteCode = routeCode,
            )
        }
        data["notificationId"]?.let { id ->
            state.value.notifications.firstOrNull { it.id == id }?.let { markNotificationRead(it) }
        }
        selectRoute(routeCode)
    }

    fun clearError() {
        _state.update { it.copy(errorMessage = null) }
    }

    fun logout(error: String? = null) {
        viewModelScope.launch {
            sessionStore.clearSessionToken()
            _state.update { AppUiState(isBootstrapping = false, errorMessage = error) }
        }
    }

    private suspend fun applySession(token: String) {
        require(token.isNotBlank()) { "The server did not return a session token." }
        sessionStore.saveSessionToken(token)
        _state.update { it.copy(sessionToken = token) }
        refreshAll()
    }

    private fun setError(error: Throwable) {
        _state.update { it.copy(isLoading = false, errorMessage = error.message ?: "Something went wrong.") }
    }
}

fun routeCodeFromText(scannedText: String): String? {
    val trimmed = scannedText.trim()
    if (trimmed.isBlank()) return null
    return routeCodeFromQuery(trimmed) ?: trimmed
}

fun routeCodeFromUri(uri: Uri): String? {
    uri.getQueryParameter("routeCode")?.let { return it }
    uri.getQueryParameter("liff.state")?.let { nested ->
        routeCodeFromText(nested)?.let { return it }
    }
    uri.getQueryParameter("sessionParams")?.let { params ->
        Regex("\"routeCode\"\\s*:\\s*\"([^\"]+)\"").find(params)?.groupValues?.getOrNull(1)?.let { return it }
    }
    return null
}

private fun routeCodeFromQuery(value: String): String? {
    val query = value.substringAfter('?', missingDelimiterValue = value)
    fun parameter(name: String): String? = query
        .split('&')
        .firstOrNull { it.substringBefore('=') == name }
        ?.substringAfter('=', "")
        ?.let { URLDecoder.decode(it, Charsets.UTF_8.name()) }
        ?.takeIf { it.isNotBlank() }

    parameter("routeCode")?.let { return it }
    parameter("liff.state")?.let { nested -> routeCodeFromText(nested)?.let { return it } }
    parameter("sessionParams")?.let { params ->
        Regex("\"routeCode\"\\s*:\\s*\"([^\"]+)\"").find(params)?.groupValues?.getOrNull(1)?.let { return it }
    }
    return null
}
